File: 0047-permutations-ii/0047-permutations-ii.cpp
class Solution {
private:
    void dfs(int n, unordered_map<int, int>& freq, 
             vector<int>& path, vector<vector<int>>& res) {
        if (path.size() == n) {
            res.push_back(path);
            return;
        }
        for (auto& [num, count] : freq) {
            if (count > 0) {
                path.push_back(num);
                freq[num]--;
                dfs(n, freq, path, res);
                path.pop_back();
                freq[num]++;
            }
        }
    }
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> path;
        unordered_map<int, int> freq;
        for (int num : nums) {
            freq[num]++;
        }        
        dfs(nums.size(), freq, path, res);
        return res;
    }
};

File: 0047-permutations-ii/0047-permutations-ii.py
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        res = []
        freq = {}
        for num in nums:
            freq[num] = freq.get(num, 0) + 1
        def dfs(path):
            if len(path) == len(nums):
                res.append(path[:])
                return
            for num in freq:
                if freq[num] > 0:
                    path.append(num)
                    freq[num] -= 1
                    dfs(path)
                    path.pop()
                    freq[num] += 1
        dfs([])
        return res